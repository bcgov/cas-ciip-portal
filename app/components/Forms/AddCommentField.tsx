import React, {useState} from 'react';
import {FieldProps} from '@rjsf/core';
import {Form, Button} from 'react-bootstrap';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faCommentAlt} from '@fortawesome/free-solid-svg-icons';
import {toast} from 'react-toastify';

const AddCommentField: React.FunctionComponent<FieldProps<string>> = (
  props
) => {
  const [expanded, setExpanded] = useState(false);
  const [value, setValue] = useState(props.formData);
  const title = props.formData ? 'Edit comment' : 'Add a comment';

  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setValue(e.target.value);
  };

  const handleCancel = () => {
    setValue(props.formData);
    setExpanded(false);
  };

  const handleSave = () => {
    props.onChange(value);
    setExpanded(false);
    toast.success(
      'Your comment was saved for review upon application submission',
      {
        autoClose: 5000,
        position: 'top-center'
      }
    );
  };

  return (
    <>
      <style jsx>{`
        .title,
        .report-field .controls {
          display: flex;
          width: 100%;
          justify-content: flex-end;
          align-items: center;
        }

        .report-field .controls {
          margin-top: 16px;
        }

        .report-field {
          background-color: #fcf8e2;
          border: 1px solid #ccc;
          border-radius: 4px;
          padding: 16px;
          margin-top: 16px;
        }
      `}</style>
      <span className="title">
        <Button variant="link" onClick={() => setExpanded(true)}>
          <FontAwesomeIcon icon={faCommentAlt} />
          &nbsp;
          {title}
        </Button>
      </span>
      {expanded && (
        <div className="report-field">
          <legend>CIIP Application Comment</legend>

          <p>
            This area is only for leaving a comment regarding the data that you
            reported on this page. This comment will be visible by the person
            reviewing this CIIP application once the application is submitted.
          </p>
          <p>
            For any technical issue, or to review documentation, please click on
            the help bubble in the bottom right of your screen.
          </p>
          <Form.Control
            as="textarea"
            rows={8}
            value={value}
            onChange={handleChange}
          />
          <div className="controls">
            <Button variant="outline-dark" onClick={handleCancel}>
              Cancel
            </Button>
            <Button
              style={{marginLeft: 16}}
              variant="success"
              onClick={handleSave}
            >
              Save Comment
            </Button>
          </div>
        </div>
      )}
    </>
  );
};

export default AddCommentField;
