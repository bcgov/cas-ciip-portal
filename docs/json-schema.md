# Usage of JSON schemas in the CIIP portal

This application's source code relies extensively on the [JSON Schema specification] to define the shape and validation of the data that is part of a CIIP application.

## Rendering of the data using @rjsf/core

In order to display the data to the user, either when displaying the forms where industry reporter input their data, or when rendering the contents of a CIIP application in a table, we rely on the [@rjsf/core] project.

TODO: explain format of objects ({schema, uiSchema, customFormats, customErrors})

### TODO: support of formulas in forms

[json schema specification]: https://json-schema.org/
[@rjsf/core]: https://github.com/rjsf-team/@rjsf/core
