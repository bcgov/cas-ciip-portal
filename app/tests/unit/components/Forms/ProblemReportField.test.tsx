import React from 'react';
import {shallow} from 'enzyme';
import AddCommentField from 'components/Forms/AddCommentField';
import {JSONSchema7TypeName} from 'json-schema';

describe('The AddCommentField component', () => {
  const defaultProps = {
    schema: {
      type: 'string' as JSONSchema7TypeName
    },
    uiSchema: {},
    idSchema: null,
    errorSchema: {},
    registry: null,
    formContext: {},
    autofocus: false,
    disabled: false,
    required: false,
    readonly: false,
    name: 'comments',
    onChange: jest.fn(),
    onBlur: jest.fn()
  };
  it('Should match the snapshot', () => {
    const wrapper = shallow(
      <AddCommentField formData="Some text" {...defaultProps} />
    );
    // Expand the field to show the text area
    wrapper.find('Button').simulate('click');

    expect(wrapper).toMatchSnapshot();
  });
});
