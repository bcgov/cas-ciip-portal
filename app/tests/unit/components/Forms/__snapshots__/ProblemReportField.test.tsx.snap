// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AddCommentField component Should match the snapshot 1`] = `
<Fragment>
  <JSXStyle
    id="2736636148"
  >
    .title.jsx-2736636148,.report-field.jsx-2736636148 .controls.jsx-2736636148{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;width:100%;-webkit-box-pack:end;-webkit-justify-content:flex-end;-ms-flex-pack:end;justify-content:flex-end;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;}.report-field.jsx-2736636148 .controls.jsx-2736636148{margin-top:16px;}.report-field.jsx-2736636148{background-color:#fcf8e2;border:1px solid #ccc;border-radius:4px;padding:16px;margin-top:16px;}
  </JSXStyle>
  <span
    className="jsx-2736636148 title"
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="link"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f27a",
              "M448 0H64C28.7 0 0 28.7 0 64v288c0 35.3 28.7 64 64 64h96v84c0 9.8 11.2 15.5 19.1 9.7L304 416h144c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64z",
            ],
            "iconName": "comment-alt",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
      Â 
      Edit comment
    </Button>
  </span>
  <div
    className="jsx-2736636148 report-field"
  >
    <legend
      className="jsx-2736636148"
    >
      CIIP Application Comment
    </legend>
    <p
      className="jsx-2736636148"
    >
      This area is only for leaving a comment regarding the data that you reported on this page. This comment will be visible by the person reviewing this CIIP application once the application is submitted.
    </p>
    <p
      className="jsx-2736636148"
    >
      For any technical issue, or to review documentation, please click on the help bubble in the bottom right of your screen.
    </p>
    <FormControl
      as="textarea"
      onChange={[Function]}
      rows={8}
      value="Some text"
    />
    <div
      className="jsx-2736636148 controls"
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="outline-dark"
      >
        Cancel
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginLeft": 16,
          }
        }
        variant="success"
      >
        Save Comment
      </Button>
    </div>
  </div>
</Fragment>
`;
